### DISCUSSION TITLE
Vertex centric index not getting used for edge property matching with previously visited edge

### DISCUSSION THREAD

**Discussion by csrajmohan** [JanusGraph/janusgraph#3056] [Q&A]
Hi, 
I am using janusgraph 0.6.1 with cassandra backend. I am facing an issue related to vertex centric indexes.

My graph is of the form:
`asset ---flow-->job ---flow-->asset 
`

and I am using the following subgraph query to get edge induced subgraph starting from a given asset vertex.
Query 1, query 2 and query 3 given below are slight variants of that query.

I have a `vertex centric index `defined on edges of type `'flow'` on the property `'flow_id'.`  
The queries have a conditional check on edge property `'flow_id'` that is matched with the `'flow_id' `property of previously visited edge in same flow. For this the vertex centric index is not getting used. 

Query 1 and query 2 are not making using of vertex centric index for conditional predicate on `flow_id` property of the edge. Here the condition is to match on the property value of already visited edge. But when I have conditional predicate on flow_id with a constant value, then it uses vertex centric index as in query 3. Why is it happening? Is there a way to rewrite the query in order to make janusgraph utilize the vertex centric index in this scenario.

Query 1:
`g.V().has("asset_id","asset1234").repeat(__.outE("flow").as("e1").subgraph("subGraph").inV().outE("flow").where(P.eq("e1")).by("flow_id").subgraph("subGraph").inV()).cap("subGraph").next()`

Query 2:
`g.V().has("asset_id","asset1234").repeat(__.outE("flow").as("e1").subgraph("subGraph").values("flow_id").as("fid").select("e1").inV().outE("flow").has("flow_id",select("fid")).subgraph("subGraph").inV()).cap("subGraph").next()`

Query 3:
`g.V().has("asset_id","asset1234").repeat(__.outE("flow").as("e1").subgraph("subGraph").values("flow_id").as("fid").select("e1").inV().outE("flow").has("flow_id", "f1").subgraph("subGraph").inV()).cap("subGraph").next()`

Thanks for the help.

**Comment by vtslab** [comment]
In the past of JanusGraph we experienced cases where indices were not hit and that could be resolved afterwards. Your use case seems important to support. Would it be possible for you to do two things:
1. explain how you conclude that the index is not hit (e.g. add profile() output)
2. rewrite your use case so that it can run on the [Graph of the Gods,](https://docs.janusgraph.org/getting-started/basic-usage/#loading-the-graph-of-the-gods-into-janusgraph) like in [https://docs.janusgraph.org/schema/index-management/index-performance/#vertex-centric-indexes](https://docs.janusgraph.org/schema/index-management/index-performance/#vertex-centric-indexes )

**Reply by csrajmohan** [reply]
@vtslab thanks for the reply.
Yes I checked profile() output for all the queries and thats how I concluded that index is used or not. Index is used only in Query 3. I can add the profile() output shortly here.

**Comment by li-boxuan** [comment]
Your query 2 and query 3 look exactly the same except they start from different vertices. Just want to make sure: did you do a reindexing after you create the index? This is required unless when you create the index, you don't have any `flow` edges yet.

**Reply by csrajmohan** [reply]
@li-boxuan Thanks for the reply. Sorry about the confusion of different starting vertices it is a typo which I fixed in the question now. All queries start at same vertex and exactly do the same thing except a small change i.e. query 3 has hard-coded string to compare while other two queries are comparing with previously read property value of an edge. Yes index is in enabled state and used in query 3. Just slight change of that in query 1 and query 2 where index is not getting used.

**Reply by li-boxuan** [reply]
Make sure your `select` returns a single value. I suspect in your query 2, `select("fid")` returns a list of values.
